### PR TITLE
Start using local address nodes in the JIT frontend

### PR DESCRIPTION
**Pull request by mikedn**
Small contribution to https://github.com/dotnet/coreclr/issues/19875 to inaugurate the new repo :)

I've been trying to figure out how to handle this, there's no way to completly remove `ADDR` in a single PR. And I think I found a reasonable good starting point - generate some `LCL_VAR|FLD_ADDR` nodes in some specific case that don't require a lot of other changes.

Of course, it would make more sense to generate these nodes directly from the importer. But that's probably going to be an all or nothing kind of thing and require far more changes.

Also, this starts another change that I was considered ever since I added `LocalAddressVisitor` - start taking over the responsibility of morphing indirect/field local variable access from global morph. The current situation is impractical, `LocalAddressVisitor` has its own analysis logic and morph has its own transformation logic. That's redundant and risky, it's hard to say if the two are in sync. Besides, global morph should have been dismantled a long time ago :)

### CODE REVIEW

**Comment by xiangzhai** [comment]
:mips-interest

**Comment by mikedn** [comment]
There are a few diffs on both xarch and arm. For example, on x64 we have:
```
Total bytes of diff: -320 (-0.00% of base)
    diff is an improvement.

Top file improvements by size (bytes):
        -200 : System.Private.CoreLib.dasm (-0.01% of base)
         -21 : System.Security.Cryptography.X509Certificates.dasm (-0.01% of base)
         -18 : xunit.assert.dasm (-0.04% of base)
         -16 : System.Net.WebSockets.dasm (-0.04% of base)
         -14 : System.Net.Http.dasm (-0.00% of base)

13 total files with size differences (13 improved, 0 regressed), 116 unchanged.

Top method regressions by size (bytes):
           3 ( 0.78% of base) : System.Private.CoreLib.dasm - BigInteger:Pow10(int,byref)

Top method improvements by size (bytes):
         -14 (-3.40% of base) : System.Private.CoreLib.dasm - Utf8Span:TryFind(Rune,byref):bool:this
         -14 (-3.40% of base) : System.Private.CoreLib.dasm - Utf8Span:TryFindLast(Rune,byref):bool:this
         -12 (-0.47% of base) : System.Security.Cryptography.X509Certificates.dasm - CertificateRequest:Create(X500DistinguishedName,X509SignatureGenerator,DateTimeOffset,DateTimeOffset,ref):X509Certificate2:this
         -11 (-3.55% of base) : System.Net.Http.dasm - TaskCompletionSourceWithCancellation`1:WaitWithCancellationAsync(CancellationToken):ValueTask`1:this (2 methods)
          -9 (-3.15% of base) : System.Console.dasm - Kernel32:GetLeadByteRanges(int,ref):int

Top method regressions by size (percentage):
           3 ( 0.78% of base) : System.Private.CoreLib.dasm - BigInteger:Pow10(int,byref)

Top method improvements by size (percentage):
          -6 (-5.94% of base) : System.Private.CoreLib.dasm - Stopwatch:QueryPerformanceFrequency():long
          -8 (-5.03% of base) : System.Private.CoreLib.dasm - Utf8String:Contains(Rune):bool:this
          -8 (-4.65% of base) : System.Private.CoreLib.dasm - Utf8Span:Contains(Rune):bool:this
          -8 (-4.47% of base) : System.Private.CoreLib.dasm - Utf8Span:StartsWith(Rune):bool:this
          -8 (-4.40% of base) : System.Private.CoreLib.dasm - ModuleHandle:GetPEKind(RuntimeModule,byref,byref)
```
These are mostly the result of local address nodes not having `GTF_GLOB_REF` (or any other side effects) which allows better call arg ordering:
```diff
--- D:\Projects\jit\diffs\dasmset_5\base\Microsoft.CodeAnalysis.CSharp.dasm	2019-11-26 23:09:41.000000000 +-0200
+++ D:\Projects\jit\diffs\dasmset_5\diff\Microsoft.CodeAnalysis.CSharp.dasm	2019-11-26 23:09:53.000000000 +-0200
@@ -468699,18 +468698,17 @@
        call     [Binder:get_Conversions():Conversions:this]
        mov      rdi, rax
-       lea      r13, bword ptr [rbp-80H]
        mov      rcx, gword ptr [rbp-B8H]
        lea      r11, [(reloc)]
        cmp      dword ptr [rcx], ecx
        call     [MethodSymbol:get_ReturnType():TypeSymbol:this]
        mov      r8, rax
        lea      rdx, bword ptr [rbp-108H]
        mov      bword ptr [rsp+20H], rdx
        xor      edx, edx
        mov      dword ptr [rsp+28H], edx
-       mov      rdx, r13
+       lea      rdx, bword ptr [rbp-80H]
        mov      rcx, rdi
        mov      r9, gword ptr [rbp-C8H]
        cmp      dword ptr [rcx], ecx
        call     [ConversionsBase:ClassifyConversion(TypeSymbol,TypeSymbol,byref,bool):Conversion:this]
        mov      rcx, rbx
```
There's actually more that can be done in this area but I removed some changes to keep the PR simpler. They can be done later as CQ improvements, independent of IR refactoring.

Another small source of diffs is the elimination of some GS shadow var copies. `gsMarkPtrsAndAssignGroups` appears to be about as bad as `fgMarkAddressExposedLocals` before I changed it and gets confused. For this tree (before this change):
```
[000107] -ACXG+------              *  JTRUE     void  
[000106] JACXG+-N----              \--*  NE        int   
[000104] -ACXG+------                 +--*  CAST      int <- bool <- int
[000101] -ACXG+------                 |  \--*  CALL r2r_ind int    System.Text.Rune.TryEncodeToUtf8
[000192] -A--------L- this SETUP      |     +--*  ASG       byref 
[000191] D------N----                 |     |  +--*  LCL_VAR   byref  V28 tmp23        
[000096] -----+------                 |     |  \--*  ADDR      byref 
[000097] ----G+-N----                 |     |     \--*  LCL_VAR   struct<System.Text.Rune, 4>(AX) V01 arg1         
[000190] -A---+----L- arg1 SETUP      |     +--*  COMMA     void  
[000183] -A----------                 |     |  +--*  COMMA     void  
...
[000188] -------N----                 |     |     \--*  LCL_VAR   int    V23 tmp18        
[000193] ------------ this in rcx     |     +--*  LCL_VAR   byref  V28 tmp23        
[000195] ------------ arg1 in rdx     |     +--*  ADDR      byref 
[000194] -------N----                 |     |  \--*  LCL_VAR   struct<System.Span`1[Byte], 16>(AX) V26 tmp21        
[000100] -----+------ arg2 in r8      |     \--*  ADDR      long  
[000099] ----G+-N----                 |        \--*  LCL_VAR   int   (AX) V12 tmp7         
[000105] -----+------                 \--*  CNS_INT   int    0
```
`gsMarkPtrsAndAssignGroups` thinks like this:
* The `this` arg of a call is considered "under indir".
* In this case the `this` arg is actually a setup assignment.
* The LHS of the assignment gets visited and thus `V28` has `lvIsPtr` set on it because it's under indir. This is a bit odd as the LHS is obviously not a real use. But `V28` will be used as a late arg so it's ok, `V28` is indeed a variable that contains a pointer.
* The RHS of the assignment is visited.
* The `ADDR` node is encountered and "under indir" is reset
* The `LCL_VAR V01` node is encountered so `lvIsPtr` is not set on `V01`. But this node is also treated as the assignment source and thus placed in the same equivalence group with `V28` which earlier was determined to be a pointer. So now GS thinks that `V01` is a pointer and makes a shadow copy of it. But it's not the value stored in `V01` that is being assigned, it's its address. `V01` itself is definitely not a pointer.

With this change the shadow copy is avoided because GS doesn't pay attention to LCL_VAR_ADDR nodes.

The new tree looks like this:
```
[000107] -ACXG+------              *  JTRUE     void  
[000106] JACXG+-N----              \--*  NE        int   
[000104] -ACXG+------                 +--*  CAST      int <- bool <- int
[000101] -ACXG+------                 |  \--*  CALL r2r_ind int    System.Text.Rune.TryEncodeToUtf8
[000192] -A--------L- this SETUP      |     +--*  ASG       byref 
[000191] D------N----                 |     |  +--*  LCL_VAR   byref  V28 tmp23        
[000096] -----+------                 |     |  \--*  LCL_VAR_ADDR byref  V01 arg1         
[000190] -A---+----L- arg1 SETUP      |     +--*  COMMA     void  
[000183] -A----------                 |     |  +--*  COMMA     void  
...
[000188] -------N----                 |     |     \--*  LCL_VAR   int    V23 tmp18        
[000193] ------------ this in rcx     |     +--*  LCL_VAR   byref  V28 tmp23        
[000195] ------------ arg1 in rdx     |     +--*  ADDR      byref 
[000194] -------N----                 |     |  \--*  LCL_VAR   struct<System.Span`1[Byte], 16>(AX) V26 tmp21        
[000100] -----+------ arg2 in r8      |     \--*  LCL_VAR_ADDR long   V12 tmp7         
[000105] -----+------                 \--*  CNS_INT   int    0
```
The `V28` is still present but as explained above, even that should go away in the future since the `LCL_VAR_ADDR` itself could be moved to become a late arg itself.

**Comment by mikedn** [comment]
There are also some text only disassembly changes caused by a `TYP_BYREF` -> `TYP_I_IMPL` change. This is due to a rationalization peculiarity - when converting `ADDR(LCL_VAR)` to `LCL_VAR_ADDR` it also changes the type of the node from `TYP_I_IMPL` to `TYP_BYREF`.

I don't know why, local variable addresses do not need to be reported to the GC. And it's not like the new code is creating local address nodes with `TYP_I_IMPL`, it simply preserves the type which at that point is `TYP_BYREF` usually.

Then `fgMorphArgs` changes the address type to `TYP_I_IMPL`:
```C++
if (argx->IsLocalAddrExpr() != nullptr)
{
    argx->gtType = TYP_I_IMPL;
}
```
only for rationalization to change it back to `TYP_BYREF`. And this no longer happens now when the node is already `LCL_VAR_ADDR`.

I suppose I could make rationalization to force `TYP_BYREF` on existing local address nodes. But that should not be needed.

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -18213,6 +18319,57 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
+    //    to a single LCL_VAR_ADDR or LCL_FLD_ADDR node.
+    //
+    // Arguments:
+    //    val - a value that represents the local address
+    //
+    void MorphLocalAddress(const Value& val)
+    {
+        assert(val.IsAddress());
+        assert(val.Node()->TypeIs(TYP_BYREF, TYP_I_IMPL));
+        assert(m_compiler->lvaVarAddrExposed(val.LclNum()));
```
Comment (CarolEidt) [inline src/coreclr/src/jit/morph.cpp]: I know that the JIT conflates the notions of a local being address-taken and address-exposed, but I think that's something we need to fix - migrating `lvHasLdAddrOp` to `lvAddrTaken` and leaving `lvAddrExposed` to actually mean that the address has potentially been exposed beyond this method. In that vein I think that this assert could allow either to be true.

File: src/coreclr/src/jit/morph.cpp [hunk 2]
```diff
@@ -18213,6 +18319,57 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
+
+        LclVarDsc* varDsc = m_compiler->lvaGetDesc(val.LclNum());
+
+        if (varDsc->lvPromoted || varDsc->lvIsStructField || m_compiler->lvaIsImplicitByRefLocal(val.LclNum()))
+        {
+            // Ignore promoted and "implict by ref" variables for now,
+            // they require additional changes in subsequent phases.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/morph.cpp]: Should this also have a "TODO"? Or perhaps instead we need a general issue to capture the various improvements that are outstanding in this area?

**Review by CarolEidt — APPROVED** [review]
This looks great overall; just a couple of questions - one about how to capture the next phase of improvements (TODOs or opening an issue) and an assert suggestion. In either case, I'm fine with things as-is, but am curious what you think.

File: src/coreclr/src/jit/morph.cpp (hunk 2 above)
Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: Yes, I think a TODO is reasonable, both here and on the other comment above.

I already added a list with the various `ADDR` use cases to the existing `ADDR` issue so it would make sense to also add a list of TODOs as well.

File: src/coreclr/src/jit/morph.cpp (hunk 1 above)
Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: > and leaving lvAddrExposed to actually mean that the address has potentially been exposed beyond this method

Hmm? While that may be useful in some cases, the JIT doesn't currently have such capability and it's not clear what it would take to add it. `LocalAddressVisitor` can detect when a local address is passed as a call argument within the same tree but it currently doesn't not track values across trees. It's likely that it can be extended to do that as well - on assignment, the  `Value` computed on the RHS could be associated with the destination local variable so we can continue tracking address uses, at least within the block if not the entire method. But such work seems rather orthogonal to `ADDR` elimination.

> I know that the JIT conflates the notions of a local being address-taken and address-exposed

The goal is to make address-taken and address-exposed equivalent:
* Any use of `LCL_VAR|FLD_ADDR` on a variable => variable is address taken/exposed and thus not tracked
* Any tree that represents an indirect access to a variable (well, any tree that `LocalAddressVisitor` can recognize as such) is converted to a single `LCL_VAR|FLD` node that represents an equivalent access. The recognized trees are expected to be a superset of the tress recognized by `IsLocalAddrExpr` & co. so this would result `IsLocalAddrExpr` no longer being necessary.
* To put it another way - liveness & SSA will only work with `LCL_VAR|FLD` and never with `LCL_VAR|FLD_ADDR`.

**Comment by sandreenko** [comment]
cc @dotnet/jit-contrib

File: src/coreclr/src/jit/morph.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/coreclr/src/jit/morph.cpp]: So ... it may be that ultimately there's no disadvantage to making these equivalent. That is, if we postpone any marking of `lvAddrExposed` until `Lowering` when an address is required for code generation, then we presumably won't unduly pessimize the handling of address-taken (as opposed to address-exposed) locals.

Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: > That is, if we postpone any marking of lvAddrExposed until Lowering when an address is required for code generation

We can't, `lvAddrExposed` has to be set before liveness in order to determine if a variable is tracked or not. In fact, it likely needs to be set even before global morph due to local assertion propagation.

>  then we presumably won't unduly pessimize the handling of address-taken (as opposed to address-exposed) locals

Ah, but there's no pessimization going on here, except perhaps the corner case of variables having field accesses with an offset larger than 65535. The "address taken but not address exposed" case does not transform into "address exposed", the case simply disappears.

Comment (CarolEidt) [reply src/coreclr/src/jit/morph.cpp]: Perhaps I wasn't clear. I believe that we want to track/optimize structs that are only address-taken due to code generation requirements where it needs the value to be on the stack, but are not actually address-exposed.
Or maybe you're thinking that for those we never mark them `lvAddrExposed`? If so, how do we indicate to the register allocator and the frame layout code that they must be on the stack?

Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: > Or maybe you're thinking that for those we never mark them lvAddrExposed?

Yes, these should not be `lvAddrExposed`.

> If so, how do we indicate to the register allocator and the frame layout code that they must be on the stack?

Isn't that what `lvDoNotEnregister` is for?

BTW, I think it is the setting of `lvDoNotEnregister` that should be postponed until lowering. Currently we're setting it as soon as we create `LCL_FLD`s in morph but I think that eventually `LCL_FLD`s should not prevent enregistration by default. But I haven't yet checked if there's anything else before register allocation that would really need `lvDoNotEnregister`.

Comment (CarolEidt) [reply src/coreclr/src/jit/morph.cpp]: Okay, that makes sense. I agree about postponing the setting of `lvDoNotEnregister` as much as possible. Although I haven't looked in detail, I believe that most of the cases where we set it are due to struct copies, argument passing and returns - and IMO we shouldn't even create `LCL_FLD`s for any of those until `Lowering` (that is, unless they're actually fields, as opposed to actually being "reinterpret casts") and it should also be responsible for determine when these contexts require the value to be on the stack.

Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: > and IMO we shouldn't even create LCL_FLDs for any of those until Lowering (that is, unless they're actually fields, as opposed to actually being "reinterpret casts") and it should also be responsible for determine when these contexts require the value to be on the stack.

Yes, I hope that this ADDR/ASG and 1st class struct saga will also lead me to deal with some reinterpret casts. And I'd also love to see most ABI handling moved to lowering, though it's not very clear to me what exactly can be moved.

There's seems to be a rather convoluted trade-off between ABI details that are not relevant to optimizations and those that are. For example - if a single float struct has to be passed in a GPR we need a `movd`. If the same struct values gets passed to 2 call arguments, does that mean that we'd want the `movd` to be created early (as `BITCAST`) so it can be picked up by CSE? Is it even worth considering such cases? While I'm convinced that the current JIT can be improved I don't think it should handle every possible case.

Speaking of variable enregistration, I have a question about LSRA: Is there anything that would force an enregistered variable to be spilled to memory? Kind of the opposite of `RegOptional`.

Comment (CarolEidt) [reply src/coreclr/src/jit/morph.cpp]: I'm not sure that the value of CSE'ing `BITCAST`s would justify supporting it. While it wouldn't be complex, it would require, as you point out, that we expose those ABI details early. That said, I think we also need to continue to consider addressing modes early on (though instead of rediscovering them in `Lowering` I'd like to see us actually create them early), which requires understanding the available addressing modes on the target. So there's clearly value in having some abstraction of some target details exposed early in the JIT. But for the ABI stuff, I think we're best off moving as much as possible to `Lowering`, and then perhaps deciding later what to reintroduce early.

> Is there anything that would force an enregistered variable to be spilled to memory? Kind of the opposite of RegOptional

If you mean something that would say "spill this to memory here", then no, there's nothing quite like that currently. The "writeThru" that I'm working on for EH-exposed variables is kind of like that, though it is applied to all defs. I also hope to build on it to enable the spilling of single-def variables while still keeping it live in a register while possible.

File: src/coreclr/src/jit/morph.cpp [hunk 3]
```diff
@@ -17624,22 +17628,90 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
+            assert(!IsLocation() && !IsAddress());
+
+            m_lclNum = lclVar->GetLclNum();
+
+            assert(m_offset == 0);
+            assert(m_fieldSeq == nullptr);
+        }
+
+        //------------------------------------------------------------------------
+        // Location: Produce an address value from a GT_LCL_VAR_ADDR node.
```
Comment (sandreenko) [inline src/coreclr/src/jit/morph.cpp]: Location -> Address.

**Review by sandreenko — COMMENTED** [review]
Good change, thank you. I am intrigued to see how that huge work item of removing `ADDR` can be split into reasonable changes.

File: src/coreclr/src/jit/morph.cpp (hunk 3 above)
Comment (mikedn) [reply src/coreclr/src/jit/morph.cpp]: Hah, I'll fix it with the next PR in this area.
